model/src/applications/callback/mod.rs

```diff
@@ -0,0 +1,135 @@
+    }
+
+    // data is intentionally not exported because it's highly likely that
+    // CommandCallbackData will not be the only additional data contained in a
+    // response.
+    fn data(&self) -> Option<&CommandCallbackData> {
+        match self {
+            InteractionResponse::ChannelMessageWithSource(d) => Some(d),
+            InteractionResponse::Pong | InteractionResponse::DeferredChannelMessageWithSource => {
+                None
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for InteractionResponse {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<InteractionResponse, D::Error> {
+        let envelope = InteractionResponseEnvelope::deserialize(deserializer)?;
+        envelope.try_into().map_err(DeError::custom)
+    }
+}
+
+impl Serialize for InteractionResponse {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        InteractionResponseEnvelope {
+            kind: self.kind(),
+            data: self.data().cloned(),
+        }
+        .serialize(serializer)
+    }
+}
+
+impl<'a> TryFrom<InteractionResponseEnvelope> for InteractionResponse {
+    type Error = InteractionResponseEnvelopeParseError;
+
+    fn try_from(envelope: InteractionResponseEnvelope) -> Result<Self, Self::Error> {
+        let i = match envelope.kind {
+            InteractionResponseType::Pong => InteractionResponse::Pong,
+            InteractionResponseType::ChannelMessageWithSource => {
+                InteractionResponse::ChannelMessageWithSource(envelope.data.ok_or(
+                    InteractionResponseEnvelopeParseError::MissingData(envelope.kind),
+                )?)
+            }
+            InteractionResponseType::DeferredChannelMessageWithSource => {
+                InteractionResponse::DeferredChannelMessageWithSource
+            }
+        };
+
+        Ok(i)
+    }
+}
+
+#[derive(Debug)]
+enum InteractionResponseEnvelopeParseError {
+    MissingData(InteractionResponseType),
+}
+
+impl Display for InteractionResponseEnvelopeParseError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self {
+            Self::MissingData(kind) => {
+                write!(f, "data not present, but required for {}", kind.kind())
+            }
+        }
+    }
+}
+
+/// Raw payload sent when responding to an interaction.
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+struct InteractionResponseEnvelope {
+    #[serde(rename = "type")]
+    pub kind: InteractionResponseType,
+    pub data: Option<CommandCallbackData>,
+}
+
+/// Contains the possible response type integers for an interaction.
+#[derive(
+    Clone, Copy, Debug, Deserialize_repr, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize_repr,
+)]
+#[repr(u8)]
+pub enum InteractionResponseType {
+    Pong = 1,
+    ChannelMessageWithSource = 4,
+    DeferredChannelMessageWithSource = 5,
+}
+
+impl InteractionResponseType {
+    pub fn kind(self) -> &'static str {
```

Comment:
We typically use `name` for these kinds of methods.

```suggestion
    pub fn name(self) -> &'static str {
```

Reply:
I think i gave the idea to that, it was because `InteractionResponse` contains a `name` field, but It will probably not be a issue when it is on the `InteractionResponseType` enum.